etl/pipeline.go

```diff
@@ -0,0 +1,313 @@
+	routes, err := newRoutes(p.transform)
+	if err != nil {
+		return nil, err
+	}
+	// First, for each output topic, we compute all of the input topics
+	// needed by all of the ETLs to that output.  Note that an input
+	// topic may be routed to multiple output topics but all of those
+	// topics must be in the same pool (XXX add a check for this).
+	for _, etl := range p.transform.ETLs {
+		switch etl.Type {
+		case "denorm":
+			if etl.Left == "" || etl.Right == "" {
+				return nil, errors.New("both 'left' and 'right' topics must be specified for denorm ETL")
+			}
+			if etl.In != "" {
+				return nil, errors.New("'in' topic cannot be specified for denorm ETL")
+			}
+			if err := routes.enter(etl.Left, etl.Out); err != nil {
+				return nil, err
+			}
+			if err := routes.enter(etl.Right, etl.Out); err != nil {
+				return nil, err
+			}
+		case "stateless":
+			if etl.In == "" {
+				return nil, errors.New("'in' topic must be specified for stateless ETL")
+			}
+			if etl.Left != "" || etl.Right != "" {
+				return nil, errors.New("'left' or 'right' topic cannot be specified for stateless ETL")
+			}
+			if err := routes.enter(etl.In, etl.Out); err != nil {
+				return nil, err
+			}
+		default:
+			return nil, fmt.Errorf("unknown ETL type: %q", etl.Type)
+		}
+	}
+	// For every input topic, we now know what output pool it is rounted to.
+	// There may be multiple output topics for a given input, but they all
+	// land in the same pool.
+
+	// For each output topic, we'll build a Zed for all the ETLs that
+	// land on that topic.   It's okay to have multiple ETLs landing
+	// on the same topic (e.g., multiple ETLs from different tables
+	// land on one denormalized table).
+
+	var zeds []string
+	for _, outputTopic := range routes.Outputs() {
+		var etls []Rule
+		for _, etl := range p.transform.ETLs {
+			if etl.Out == outputTopic {
+				etls = append(etls, etl)
+			}
+		}
+		inputTopics := routes.InputsOf(outputTopic)
+		s, err := buildZed(inputTopics, outputTopic, routes, etls)
+		if err != nil {
+			return nil, err
+		}
+		zeds = append(zeds, s)
+	}
+	return zeds, nil
+}
+
+//XXX TBD: this is currently taking all records for an update in memory.
+// We need to work out a streaming model so we can do a very large updates
+// that could involve large anti-joins and spills and we stream the output
+// to the target pool.  This will scale fine since the anti-join design
+// allows any commits to be reliably accounted for on an interruption and restart.
+func (p *Pipeline) writeToOutputPool(ctx context.Context, batch zbuf.Array) error {
+	offsets, err := p.outputPool.NextProducerOffsets()
+	if err != nil {
+		return err
+	}
+	out, err := insertOffsets(ctx, p.zctx, p.doneType, batch, offsets)
+	if err != nil {
+		return err
+	}
+	vals := batch.Values()
+	for k := range vals {
+		// This pointer comparison should work but it doesn't right now.
+		// Are they all allocated in the same zctx?
+		//if vals[k].Type == p.doneType {
```

Comment:
They aren't!

Reply:
I'm going to leave this commented out for now so I don't have to retest since the existing logic works.  This is all a moving target so I don't want to waste too much time finessing things that will change.

Reply:
Actually, I tried fixing and reran the demo but it still doesn't work.  I created issue #31 31